File: backend/app/integrations/inngest.py
"""Thin Inngest event-send helper for the Python worker.

The worker fires events into Inngest whenever a provider webhook arrives
(Zoom/Teams/Recall). Inngest's ingestion API is a simple JSON POST to
``https://inn.gs/e/{event_key}`` — no Python SDK needed.

Docs: https://www.inngest.com/docs/events
"""

from __future__ import annotations

import logging
from typing import Any

import httpx

from app.core.config import settings

logger = logging.getLogger(__name__)

INNGEST_INGEST_URL = "https://inn.gs/e"


async def send_event(name: str, data: dict[str, Any]) -> None:
    """Fire an event into Inngest. No-op if INNGEST_EVENT_KEY is unset.

    Failures are logged but never raised — webhook endpoints should still
    200 back to the provider so they don't retry forever on transient
    Inngest outages. Inngest has its own retry/durability; we're just
    responsible for getting the event to them once.
    """
    if not settings.inngest_event_key:
        logger.warning(
            "inngest_event_dropped_no_key event=%s (set INNGEST_EVENT_KEY)", name
        )
        return

    payload = {"name": name, "data": data}
    try:
        async with httpx.AsyncClient(timeout=5) as client:
            resp = await client.post(
                f"{INNGEST_INGEST_URL}/{settings.inngest_event_key}",
                json=payload,
            )
            if resp.status_code >= 400:
                logger.error(
                    "inngest_event_send_failed event=%s status=%d body=%s",
                    name,
                    resp.status_code,
                    resp.text[:300],
                )
                return
            logger.info("inngest_event_sent event=%s", name)
    except Exception as exc:  # noqa: BLE001
        logger.exception("inngest_event_send_error event=%s error=%s", name, exc)

File: worker/internal/integrations/inngest/inngest.go
// Package inngest fires events into Inngest's ingestion API (a plain JSON POST
// to https://inn.gs/e/{event_key}) — ports app/integrations/inngest.py. The
// worker fires events when a provider webhook arrives or a bot finishes; the
// Inngest functions (TypeScript on Vercel) orchestrate the pipeline from there.
package inngest

import (
	"bytes"
	"context"
	"encoding/json"
	"log/slog"
	"net/http"
	"time"
)

const ingestBaseURL = "https://inn.gs/e"

// Sender posts events. A zero EventKey makes Send a logged no-op (matching the
// Python "no key -> drop" behaviour) so webhook handlers can still ack.
type Sender struct {
	EventKey   string
	HTTPClient *http.Client
}

// New builds a Sender with a 5s timeout (webhook handlers must ack quickly).
func New(eventKey string) *Sender {
	return &Sender{EventKey: eventKey, HTTPClient: &http.Client{Timeout: 5 * time.Second}}
}

// Send fires one event. Failures are logged, never returned as fatal — the
// caller (a webhook handler) should still 200 the provider so it doesn't retry
// forever; Inngest has its own durability for events that do land.
func (s *Sender) Send(ctx context.Context, name string, data map[string]any) {
	if s.EventKey == "" {
		slog.Warn("inngest event dropped: no key", "event", name)
		return
	}
	payload, err := json.Marshal(map[string]any{"name": name, "data": data})
	if err != nil {
		slog.Error("inngest event marshal failed", "event", name, "err", err)
		return
	}
	hc := s.HTTPClient
	if hc == nil {
		hc = &http.Client{Timeout: 5 * time.Second}
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, ingestBaseURL+"/"+s.EventKey, bytes.NewReader(payload))
	if err != nil {
		slog.Error("inngest event request failed", "event", name, "err", err)
		return
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := hc.Do(req)
	if err != nil {
		slog.Error("inngest event send error", "event", name, "err", err)
		return
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		slog.Error("inngest event send failed", "event", name, "status", resp.StatusCode)
		return
	}
	slog.Info("inngest event sent", "event", name)
}
